File: _CodeTopics/LeetCode/000207/000207_ugly.py
class Solution(object):
    hasRing = False
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """

        visited = [False for i in range(numCourses)]
        # 后来发现不注释这个定义，然后下面dfs函数里用global hasRing还是不行，只好先用类成员变量了。
        # hasRing = False

        # 构造邻接表
        # edges = {}
        # for elem in prerequisites:
        #     if not edges.has_key(elem[1]):
        #         edges[elem[1]] = [elem[0]]
        #     else:
        #         edges[elem[1]].append(elem[0])
        # print edges

        edges = collections.defaultdict(list)
        for info in prerequisites:
            edges[info[1]].append(info[0])
        ##print edges

        def dfs(vertex):
            # python2没有nonlocal关键词，只好用global了。当然也可以把这个变量定义成类的
            # 一个成员，但是那样更丑了其实。
            # global hasRing
            visited[vertex] = "ongoing"
            for v in edges[vertex]:
                if visited[v] == False:
                    dfs(v)
                    if self.hasRing:
                        return
                elif visited[v] == "ongoing":
                    self.hasRing = True
                    return
            visited[vertex] = True

        for i in range(numCourses):
            if not self.hasRing and not visited[i]:
                dfs(i)
        return not self.hasRing
        
"""
https://leetcode-cn.com/submissions/detail/94672410/

46 / 46 个通过测试用例
状态：通过
执行用时：32 ms
内存消耗：15.8 MB
"""
"""
执行用时：32 ms, 在所有 Python 提交中击败了65.23%的用户
内存消耗：15.8 MB, 在所有 Python 提交中击败了5.55%的用户
"""

File: _CodeTopics/LeetCode/201-400/000207/tran/000207_ugly.cpp
// https://leetcode.cn/problems/median-of-two-sorted-arrays/submissions/316879101/
class Solution {
public:
    bool hasRing = false;
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> visited(numCourses, 0);  // 0 表示未访问过的节点
        unordered_map<int, vector<int>> edges;
        for (vector<int> info : prerequisites) {
            edges[info[1]].push_back(info[0]);
        }
        for (int i = 0; i < numCourses; i++) {
            if (!hasRing && visited[i] == 0) {
                dfs(i, visited, edges);
            }
        }
        return !hasRing;
    }
private:
    void dfs(int vertex, vector<int>& visited, unordered_map<int, vector<int>>& edges) {
        visited[vertex] = 1;  // 1 表示正在访问的节点
        for (int v : edges[vertex]) {
            if (visited[v] == 0) {
                dfs(v, visited, edges);
                if (hasRing) {
                    return;
                }
            }
            else if (visited[v] == 1) {
                hasRing = true;
                return;
            }
        }
        visited[vertex] = 2;  // 2 表示已经访问结束的节点
    }
};

/*
https://leetcode.cn/problems/course-schedule/submissions/464653813/?envType=daily-question&envId=2023-09-09

时间
详情
24ms
击败 40.03%使用 C++ 的用户
内存
详情
14.70MB
击败 8.17%使用 C++ 的用户
*/
